File: homework1/src/warmup.js
function change(amt) {
  if (amt < 0) { throw new RangeError(); }
  const result = [];
  let remaining = amt;
  [25, 10, 5, 1].forEach((coin) => {
    result.push(Math.floor(remaining / coin));
    remaining %= coin;
  });
  return result;
}

function stripQuotes(str) {
  let result = '';
  for (let i = 0; i < str.length; i += 1) {
    if (str.charAt(i) !== '"' && str.charAt(i) !== '\'') {
      result += str.charAt(i);
    }
  }
  return result;
}

function scramble(str) {
  let result = '';
  const charArr = str.split('');
  charArr.forEach((char) => {
    const random = Math.floor(Math.random() * str.length);
    [charArr[str.indexOf(char)], charArr[random]] =
    [charArr[random], charArr[str.indexOf(char)]];
  });
  for (let i = charArr.length; i > 0; i -= 1) {
    const random = Math.floor(Math.random() * i);
    [charArr[i - 1], charArr[random]] = [charArr[random], charArr[i - 1]];
  }
  charArr.forEach((char) => { result += char; });
  return result;
}

function powers(base, lim, callback) {
  for (let currentPow = 1; currentPow <= lim; currentPow *= base) {
    callback(currentPow);
  }
}

function* powersGenerator(base, lim) {
  let currentPow = 1;
  while (currentPow <= lim) {
    yield currentPow;
    currentPow *= base;
  }
}

function say(str) {
  if (str === undefined) { return ''; }
  let result = `${str} `;
  const nextWord = (next) => {
    if (next !== undefined) {
      result += `${next} `;
      return nextWord;
    }
    return result.trim();
  };
  return nextWord;
}

function interleave(arr, ...args) {
  const a = arr.slice();
  const b = args.slice();
  const result = [];
  while (a.length > 0 || b.length > 0) {
    if (a[0] !== undefined) { result.push(a.shift()); }
    if (b[0] !== undefined) { result.push(b.shift()); }
  }
  return result;
}

function cylinder(spec) {
  let { radius = 1, height = 1 } = spec;
  const volume = () => Math.PI * radius * radius * height;
  const surfaceArea = () =>
    (2 * Math.PI * radius * height) + (2 * Math.PI * radius * radius);
  const widen = (factor) => { radius *= factor; };
  const stretch = (factor) => { height *= factor; };

  const result = {
    volume, surfaceArea, widen, stretch,
  };

  Object.defineProperty(result, 'radius', { get: () => radius });
  Object.defineProperty(result, 'height', { get: () => height });

  return Object.freeze(result);
}

const crypto = require('crypto');

function makeCryptoFunctions(key, alg) {
  const encrypt = (str) => {
    const cipher = crypto.createCipher(alg, key);
    let crypted = cipher.update(str, 'utf8', 'hex');
    crypted += cipher.final('hex');
    return crypted;
  };
  const decrypt = (str) => {
    const decipher = crypto.createDecipher(alg, key);
    let decrypted = decipher.update(str, 'hex', 'utf8');
    decrypted += decipher.final('utf8');
    return decrypted;
  };
  return [encrypt, decrypt];
}

const rp = require('request-promise');

function randomName(spec) {
  const options = {
    uri: 'http://uinames.com/api/',
    qs: {
      amount: 1,
      gender: spec.gender,
      region: spec.region,
    },
    json: true,
  };
  return rp(options).then(response => `${response.surname}, ${response.name}`)
    .catch(err => err.status);
}

module.exports = {
  change,
  stripQuotes,
  scramble,
  powers,
  powersGenerator,
  say,
  interleave,
  cylinder,
  makeCryptoFunctions,
  randomName,
};

File: homework2/warmup.py
# pragma pylint: disable=E0001
# pragma pylint: disable=C0111
# pragma pylint: disable=C0103

from random import shuffle
import math
from itertools import product
import requests

def change(amount):
    if amount < 0:
        raise ValueError('amount cannot be negative')
    result = []
    for c in (25, 10, 5, 1):
        result += [amount // c]
        amount %= c
    return tuple(result)

def strip_quotes(string):
    return string.replace('\'', '').replace('"', '')

def scramble(string):
    result = list(string)
    shuffle(result)
    return ''.join(result)

def powers(base, limit):
    current = 1
    while current <= limit:
        yield current
        current *= base

def triples(largest_hypotenuse):
    result = []
    pairs = list(product(range(1, largest_hypotenuse), repeat=2))
    for i, j in pairs:
        if (j, i) in pairs:
            pairs.remove((j, i))
    for a, b in pairs:
        c = math.sqrt(a * a + b * b)
        if c % 1 == 0 and c <= largest_hypotenuse:
            result += [(a if a < b else b, b if b > a else a, int(c))]
    return result

def say(string1=None):

    if string1 is None:
        return ''
    def _inner_say(string2=None):
        """
        if val is None we return _inner_adder.v
        else we increment and return ourselves
        """
        if string2 is None:
            return _inner_say.string1
        _inner_say.string1 += ' ' + string2
        return _inner_say
    _inner_say.string1 = string1  # save value
    return _inner_say

def interleave():
    return 1

class Cylinder(object):
    def __init__(self, radius=1, height=1):
        self.radius = radius
        self.height = height

    @property
    def surface_area(self):
        return 2 * math.pi * (self.radius * self.radius + self.radius * self.height)

    @property
    def volume(self):
        return self.radius * self.radius * math.pi * self.height

    def widen(self, factor):
        self.radius *= factor

    def stretch(self, factor):
        self.height *= factor

def make_crypto_functions():
    return 1

def random_name(gender, region):
    response = requests.get('http://uinames.com/api', \
        params={'gender': gender, 'region': region, 'amount': '1'})
    if response.status_code == 400:
        raise ValueError(response.text)
    person = response.json()
    return '{}, {}'.format(person['surname'], person['name'])
